_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+defines the type and cannot be shadowed.
+
+### The Future of User-Defined Operators and UDVTs
+Now that we know the new feature and its limitations, the  question on everyone's mind is probably
+"Will it stay that way?".
+How will the feature evolve in the short and long term?
+
+#### Literal Suffixes
+The very next step for UDVTs is the ability to provide custom literals of user-defined value types.
+The ease of using literal values in expressions is a big part of what makes a type feel native.
+Solidity already provides fractional literals and our intention is to make them usable with custom
+fixed-point number implementations.
+To achieve that we are going to allow defining suffix functions that return UDVT by converting
+literals that are already available in the language:
+
+```solidity
+type UFixed16x2 is uint16;
+
+function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
+```
+```solidity
+UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
+```
+
+While the application of a suffix will involve a function call, eventually, with the help of
+compile-time evaluation, literal suffixes are going to become as efficient as native literals.
+
+#### More Operators
+Many operators have been excluded from the initial implementation.
+Some of them will very likely be user-definable at some point.
+
+1. `<<`, `>>` and `**` were excluded due to their non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1.  `!` was excluded because we have to decide whether it should return the same type as the argument
+    of just `bool`.
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
```

Comment:
This is a *lot* of conjecture again... not sure we'll have a spaceship operators, not sure we *won't* have boolean conjunction and disjunction... does any of this information really help?

Also the last enumeration has two ``1.``s :-).

Reply:
Well, it's worded hypothetically. The conjecture is the point. It's just a blog post, we can afford to speculate. It's not like we're making binding promises here.

I think that talking about these things more in the open will help get us more feedback. It puts the ideas out here, people can react to them.

Reply:
So, is any of this actually wrong (not just unlikely?). Or is there are more likely alternative. Feel free to rewrite bits that you think are too far out there. Honestly, it's also a way for me to get these ideas out of you and @chriseth :)

Reply:
Just remove it, there's no value in engaging in hypotheticals here.

Reply:
And yes, we should have and should talk about this stuff more up front, but let's keep the post clean of hypotheticals.

Reply:
The whole section about the future of operators removed.